iso15118/secc/controller/evse_data.py

```diff
@@ -0,0 +1,143 @@
+    max_power_asymmetry: Optional[float] = None
+    evse_power_ramp_limit: Optional[float] = None
+
+    evse_present_active_power: Optional[float] = None  # Optional in AC Scheduled CL
+    evse_present_active_power_l2: Optional[float] = None  # Optional in AC Scheduled CL
+    evse_present_active_power_l3: Optional[float] = None  # Optional in AC Scheduled CL
+
+
+@dataclass
+class ACBPTLimits(Limits):
+    evse_max_discharge_power: Optional[float] = None  # Required
+    evse_max_discharge_power_l2: Optional[float] = None  # Optional
+    evse_max_discharge_power_l3: Optional[float] = None  # Optional
+
+    evse_min_discharge_power: Optional[float] = None  # Required
+    evse_min_discharge_power_l2: Optional[float] = None  # Optional
+    evse_min_discharge_power_l3: Optional[float] = None  # Optional
+
+
+@dataclass
+class DCCLLimits(Limits):
+    # Optional in 15118-20 DC Scheduled CL
+    evse_max_charge_power: Optional[float] = None  # Required in 15118-20 Dynamic CL
+    evse_min_charge_power: Optional[float] = None  # Required in 15118-20 Dynamic CL
+    evse_max_charge_current: Optional[float] = None  # Required in 15118-20 Dynamic CL
+    evse_max_voltage: Optional[float] = None  # Required in 15118-20 Dynamic CL
+
+    # Optional and present in 15118-20 DC BPT CL (Scheduled)
+    evse_max_discharge_power: Optional[float] = None  # Req in 15118-20 Dynamic BPT CL
+    evse_min_discharge_power: Optional[float] = None  # Req in 15118-20 Dynamic BPT CL
+    evse_max_discharge_current: Optional[float] = None  # Req in 15118-20 Dynamic BPT CL
+    evse_min_voltage: Optional[float] = None  # Required in 15118-20 Dynamic BPT CL
+
+
+@dataclass
+class DCBPTLimits(Limits):
+    # Required in 15118-20 DC BPT CPD
+    evse_max_discharge_power: Optional[float] = None
+    evse_min_discharge_power: Optional[float] = None
+    evse_max_discharge_current: Optional[float] = None
+    evse_min_discharge_current: Optional[float] = None
+
+
+@dataclass
+class DCLimits(Limits):
+    # Required in 15118-20 DC CPD
+    evse_max_charge_power: Optional[float] = None  # Required for -2 DC, DIN CPD
+    evse_min_charge_power: Optional[float] = None  # Not Required for -2 DC, DIN CPD
+    evse_max_charge_current: Optional[float] = None  # Required for -2 DC, DIN CPD
+    evse_min_charge_current: Optional[float] = None  # Required for -2 DC, DIN CPD
+    evse_max_voltage: Optional[float] = None  # Required for -2 DC, DIN CPD
+    evse_min_voltage: Optional[float] = None  # Required for -2 DC, DIN CPD
+
+    #  Optional in 15118-20 DC CPD
+    evse_power_ramp_limit: Optional[float] = None
+
+    #  Optional in 15118-2 CPD
+    evse_current_regulation_tolerance: Optional[float] = None
+    evse_peak_current_ripple: Optional[float] = None
+    evse_energy_to_be_delivered: Optional[float] = None
+
+
+@dataclass
+class EVSERatedLimits:
+    ac_limits: Optional[ACLimits] = None
+    ac_bpt_limits: Optional[ACBPTLimits] = None
```

Comment:
My thinking was as cs limits splits these, this would make it easier for implementers. This would make it evident that the bpt bit could be skipped if not interesting.

Reply:
Yes, if picking the bpt version, the non-bpt equivalent will also be required.
We can't make ac_limits non optional here as dc_limits also lives in the same RatedLimits dataclass.
One option is to let the bpt limits live as optional ones in the ac_limits/dc_limits dataclass. But splitting them makes it slightly easier to read I think. But as these rated limits are built from cs-config we can clarify (will have to be in the documentation though) that this is needed. Hmm, do we need add a __post_init__() to validate?